File: Team Membership Tasks/Task7-StateMachine/State Machine Implementation.py
import numpy as np

m = 8
def addarr(arr, n, g1, g2):
    for i in range(n):
        for j in range(n):
            arr[i][j] = 0
    arr[g1][g2] = 1

def showarr(arr, n):
    for i in range(n):
        for j in range(n):
            print(f"[{i}] [{j}]= {arr[i][j]}\t", end="")
        print("\n")

def move_u(arr, n, g1, g2):
    x = int(input("How many moves?\n"))
    if x > g1:
        print("Can not move!!\n")
    else:
        arr[g1][g2] = 0
        g1 -= x
        arr[g1][g2] = 1
        showarr(arr, n)

def move_d(arr, n, g1, g2):
    x = int(input("How many moves?\n"))
    if x + g1 > 7:
        print("Can not move!!\n")
    else:
        arr[g1][g2] = 0
        g1 += x
        arr[g1][g2] = 1
        showarr(arr, n)

def move_l(arr, n, g1, g2):
    x = int(input("How many moves?\n"))
    if x > g2:
        print("Can not move!!\n")
    else:
        arr[g1][g2] = 0
        g2 -= x
        arr[g1][g2] = 1
        showarr(arr, n)

def move_r(arr, n, g1, g2):
    x = int(input("How many moves?\n"))
    if x + g2 > 7:
        print("Can not move!!\n")
    else:
        arr[g1][g2] = 0
        g2 += x
        arr[g1][g2] = 1
        showarr(arr, n)

arr = np.zeros((8, 8), dtype=int)
g1, g2 = 7, 0
n = arr.shape[0]

addarr(arr, n, g1, g2)
showarr(arr, n)

while True:
    x = input("U or D or L or R or UL=i or UR=o or DL=p or DR=k (exit = E)\n")
    if x in ['U', 'u']:
        move_u(arr, n, g1, g2)
    elif x in ['D', 'd']:
        move_d(arr, n, g1, g2)
    elif x in ['L', 'l']:
        move_l(arr, n, g1, g2)
    elif x in ['R', 'r']:
        move_r(arr, n, g1, g2)
    elif x in ['E', 'e']:
        break
    else:
        print("\nU or D or L or R or UL=i or UR=o or DL=p or DR=k (exit = E)\n")

File: Team Membership Tasks/Task7-StateMachine/State Machine Implementation.cpp
#include <iostream>

using namespace std;

const int m = 8;

// Function to initialize the array with a single 1 at a given position
void initializeArray(int (*arr)[m], int n, int& row, int& col) {
   for (int i = 0; i < n; i++) {
       for (int j = 0; j < n; j++) {
           arr[i][j] = 0;
       }
   }
   arr[row][col] = 1;
}

// Function to display the array
void displayArray(int (*arr)[m], int n) {
   for (int i = 0; i < n; i++) {
       for (int j = 0; j < n; j++) {
           cout << "[" << i << "][" << j << "]= " << arr[i][j] << "\t";
       }
       cout << endl << endl;
   }
}

// Function to move the 1 up
void moveUp(int (*arr)[m], int n, int& row, int& col) {
   int moves;
   cout << "How many moves? ";
   cin >> moves;

   if (moves > row) {
       cout << "Cannot move!!" << endl << endl;
   } else {
       arr[row][col] = 0;
       row -= moves;
       arr[row][col] = 1;
       displayArray(arr, n);
   }
}

// Function to move the 1 down
void moveDown(int (*arr)[m], int n, int& row, int& col) {
   int moves;
   cout << "How many moves? ";
   cin >> moves;

   if (row + moves >= n) {
       cout << "Cannot move!!" << endl << endl;
   } else {
       arr[row][col] = 0;
       row += moves;
       arr[row][col] = 1;
       displayArray(arr, n);
   }
}

// Function to move the 1 left
void moveLeft(int (*arr)[m], int n, int& row, int& col) {
   int moves;
   cout << "How many moves? ";
   cin >> moves;

   if (moves > col) {
       cout << "Cannot move!!" << endl << endl;
   } else {
       arr[row][col] = 0;
       col -= moves;
       arr[row][col] = 1;
       displayArray(arr, n);
   }
}

// Function to move the 1 right
void moveRight(int (*arr)[m], int n, int& row, int& col) {
   int moves;
   cout << "How many moves? ";
   cin >> moves;

   if (col + moves >= n) {
       cout << "Cannot move!!" << endl << endl;
   } else {
       arr[row][col] = 0;
       col += moves;
       arr[row][col] = 1;
       displayArray(arr, n);
   }
}

int main() {
   int arr[m][m];
   int n = sizeof(arr) / sizeof(arr[0]);
   int row = 7, col = 0;
   char choice;

   initializeArray(arr, n, row, col);
   displayArray(arr, n);

   while (true) {
       cout << "U(Up) | D(Down) | L(Left) | R(Right) | UL(UpperLeft) | UR(UpperRight) | DL(DownLeft) | DR(DownRight) | E(Exit)" << endl;
       cin >> choice;

       switch (toupper(choice)) {
           case 'U':
               moveUp(arr, n, row, col);
               break;
           case 'D':
               moveDown(arr, n, row, col);
               break;
           case 'L':
               moveLeft(arr, n, row, col);
               break;
           case 'R':
               moveRight(arr, n, row, col);
               break;
           case 'E':
               return 0;
           default:
               cout << "Invalid choice! Please try again." << endl;
               break;
       }
   }

   return 0;
}
